content/posts/sieve.md

```diff
@@ -9,4 +9,74 @@ author: Bryan Li
+
+This paper introduces and analyzes the SIEVE algorithm, a simple but high-efficiency variation on the FIFO-reinsertion algorithm optimized for web cache workloads.
+
+## Background
+
+**Metrics:** There are two primary metrics used to measure cache performance: efficiency, throughput performance, and scalability. Generally, as complexity increases, throughput performance and scalability decreases as longer critical sections and more metadata may be necessary.   <!--Cache efficiency metrics include object miss ratio (fraction of requests that are cache misses) and byte miss ratio (fraction of bytes that are cache misses). Throughput performance -->
+
+**Access Patterns:** Web cache workloads typically follow a power-law (Zipfian) distribution, where a small portion of objects make up the majority of the requests. An important distinction from other cache use cases is that scan and loop/sequential access patterns are rare, and new objects are extremely common, being created every second.
+
+To expand on the power-law distribution, it follows a pattern where the $i^{th}$ most popular object has a relative frequency of $1/i^\alpha$. $\alpha$ here is a parameter determining skewness and has been found to range from 0.55 to 1.5 depending on type of workload, cache level, etc.
+
+**Efficient Cache Eviction Algorithms:** Many of the recent algorithms have greatly increased in complexity, using multiple LRU queues, machine learning, and so on. However, though they might perform better in certain workloads, the improvement is often not significant. In this way the increase in complexity outweighs the benefits.
+
+What makes a cache eviction algorithm more efficient can be boiled down into two parts: **lazy promotion** and **quick demotion**.
+
+- Lazy promotion promotes cached objects only when they are about to be evicted, retaining popular objects with minimal effort.
+  - For comparison, FIFO has no promotion, and LRU promotes on cache hit (eager promotion).
+- Quick demotion removes most objects shortly after they are cached, which takes advantage of the characteristics of the power-law distribution - most objects are unpopular.
+
+## Algorithm
+
+### Overview
+
+SIEVE uses a single FIFO queue and "hand" pointer.
+
+The queue maintains insertion order, and new objects are added to the head of the queue.
+
+Every object in the queue uses 1 bit to track visited/non-visited status, with the hand pointing to the next eviction candidate. The hand pointer moves from the tail to the head.
+
+- **Cache hits** set the visited bit of the cache line to 1.
+  - This also happens to be lock-free, which improves throughput.
+- **Cache misses** look at the head pointer and set the visited bit to 0 if it is 1, moving the hand pointer.
+  - Otherwise (`visited=0`) it will remove the eviction candidate from the cache, and move the hand pointer.
+
```

Comment:
```suggestion
- The object is added to the head of the queue with bit set to 0.
```

I know this is mentioned above, but might be nice to specify that the bit is set to 0 initially?